File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        current = head # <= 0  1 => 2 => 3 => None
        while(current):
            temp = current.next # 1, 2,
            current.next = prev # None
            prev = current # 0
            current = temp # 1
        return prev
        

        

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        if (head == nullptr || head->next == nullptr) return false;
        ListNode* slowPointer = head;
        ListNode* fastPointer = head;
        while(fastPointer != nullptr && fastPointer->next !=nullptr){
            slowPointer = slowPointer->next;
            fastPointer = fastPointer->next->next;
            if (fastPointer == slowPointer) return true;
        }
        return false;
    }
};
